### PR TITLE
ASoC: Intel: sof_rt1308: support cs35l41 amplifier

### PR DESCRIPTION
**Pull request by brentlu**
This machine driver is for boards which do not implement headphone
codec. The board config adl_cs35l41 supports Cirrus Logic
CS35L41/CS35L53 amplifier running on ADL platform.

### CODE REVIEW

**Comment by brentlu** [comment]
SOF PR: https://github.com/thesofproject/sof/pull/5372

File: sound/soc/intel/boards/sof_amp_only.c [hunk 1]
```diff
@@ -0,0 +1,552 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright(c) 2021 Intel Corporation.
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_amp_only.c]: 2022

File: sound/soc/intel/boards/sof_amp_only.c [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+	if (!hdmi_num)
+		hdmi_num = 3;
+
+	dev_dbg(&pdev->dev, "sof_amp_only_quirk = %lx\n", sof_amp_only_quirk);
+
+	/* compute number of dai links */
+	sof_amp_only_audio_card.num_links = dmic_be_num + hdmi_num;
+
+	if (sof_amp_only_quirk & SOF_SPEAKER_AMP_PRESENT)
+		sof_amp_only_audio_card.num_links++;
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_amp_only.c]: Can we assume that speaker amp is always present?

**Review by bardliao — COMMENTED** [review]
@brentlu Can we squash the three commits into one commit?

File: sound/soc/intel/boards/sof_amp_only.c [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+		.pm = &snd_soc_pm_ops,
+	},
+	.id_table = board_ids,
+};
+module_platform_driver(sof_amp_only_audio_driver)
+
+/* Module information */
+MODULE_DESCRIPTION("SOF Audio Machine driver for amplifier only boards");
+MODULE_AUTHOR("Brent Lu <brent.lu@intel.com>");
+MODULE_LICENSE("GPL v2");
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_amp_only.c]: "GPL"

File: sound/soc/intel/boards/Kconfig [hunk 4]
```diff
@@ -500,6 +500,21 @@ config SND_SOC_INTEL_SOF_CS42L42_MACH
...
+	select SND_SOC_CS35L41_I2C
+	select SND_SOC_DMIC
+	select SND_SOC_HDAC_HDMI
+	select SND_SOC_INTEL_HDA_DSP_COMMON
+	help
+	   This adds support for ASoC machine driver for SOF platforms
+	   with speaker amplifier only.
+	   Say Y if you have such a device.
+	   If unsure select "N".
+
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig]: I wonder if we could reuse the work already done for sof_rt1308.c, where there's also no headset codec.

It's the same case of an amplifier, the differences are that you have BT offload and they have HDMI inputs.

Still very very similar so I think there is merit in exploring reuse.

@CBala21 @kv2019i @ujfalusi @bardliao what do you think?

**Review by plbossart — COMMENTED** [review]
The code looks alright with minor changes, I am not sure though if it makes sense for Intel to provide two separate machine drivers for a very similar configuration, see below.

File: sound/soc/intel/boards/Kconfig (hunk 4 above)
Comment (CBala21) [reply sound/soc/intel/boards/Kconfig]: I'm not sure whether its ok to mix Cirrus and Realtek in single driver. will it cause any conflict in selection of kconfig?
Also here HDAC HDMI support added but its not present on RT1308. anyhow its not big issue to add top on it.

Comment (bardliao) [reply sound/soc/intel/boards/Kconfig]: > I'm not sure whether its ok to mix Cirrus and Realtek in single driver. will it cause any conflict in selection of kconfig? 

That should be fine. We already mix Maxim and Realtek codecs in [sof_sdw.c](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw.c)

Comment (plbossart) [reply sound/soc/intel/boards/Kconfig]: I think @CBala21 's concern was if they only want to add the Realtek RT1308 dependency. That could be done with two entries in the Kconfig selecting common code where we have a IS_ENABLED() test. We've done this in the past with the DA7219 stuff selecting common parts in this folder already.

My point is really that more than half of the code is common.

Comment (brentlu) [reply sound/soc/intel/boards/Kconfig]: I'll try to add cs35l41 support to the rt1308 machine driver and have ODM to test it.

File: sound/soc/intel/boards/Kconfig [hunk 5]
```diff
@@ -611,11 +611,15 @@ config SND_SOC_INTEL_SOF_RT1308_MACH
...
 	select SND_SOC_DMIC
+	select SND_SOC_HDAC_HDMI
+	select SND_SOC_INTEL_HDA_DSP_COMMON
 	select SND_SOC_INTEL_SOF_REALTEK_COMMON
+	select SND_SOC_INTEL_SOF_CIRRUS_COMMON
 	help
-	   This adds support for ASoC machine driver for Tigerlake platforms
-	   with RT1308 I2S audio codec.
+	   This adds support for ASoC machine driver for SOF platforms
+	   with RT1308/CS35L41 I2S audio codec.
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig:622]: also need to update the line 609 and 610

File: sound/soc/intel/boards/sof_rt1308.c [hunk 6]
```diff
@@ -178,16 +286,96 @@ static struct snd_soc_dai_link *sof_card_dai_links_create(struct device *dev,
 		id++;
 	}
 
+	/* HDMI-Out SSP */
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt1308.c]: no, this is not based on SSP (unlike HDMI-in).

Suggest renaming as /* HDMI playback */

File: sound/soc/intel/boards/sof_rt1308.c [hunk 7]
```diff
@@ -236,6 +449,15 @@ static const struct platform_device_id board_ids[] = {
 					SOF_SSP_HDMI_CAPTURE_PRESENT |
 					SOF_RT1308_SPEAKER_AMP_PRESENT),
 	},
+	{
+		.name = "adl_cs35l41",
+		.driver_data = (kernel_ulong_t)(SOF_RT1308_SSP_CODEC(1) |
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt1308.c]: you don't have an RT1308 on this platform.

File: sound/soc/intel/boards/sof_rt1308.c [hunk 8]
```diff
@@ -236,6 +449,15 @@ static const struct platform_device_id board_ids[] = {
...
+					SOF_NO_OF_HDMI_PLAYBACK_SSP(4) |
+					SOF_SSP_HDMI_PLAYBACK_PRESENT |
+					SOF_BT_OFFLOAD_SSP(2) |
+					SOF_SSP_BT_OFFLOAD_PRESENT |
+					SOF_CS35L41_SPEAKER_AMP_PRESENT),
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt1308.c]: you may be missing a macro similar to SOF_RT1308_SSP_CODEC()?

**Review by plbossart — CHANGES_REQUESTED** [review]
That sounds really good to me @brentlu, this will also give @CBala21 's customers the ability to switch amplifiers if they want to.

I only have a couple of comments but one more round and this should be good to go.

File: sound/soc/intel/boards/sof_rt1308.c (hunk 6 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_rt1308.c]: fixed

File: sound/soc/intel/boards/sof_rt1308.c (hunk 7 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_rt1308.c]: The macro SOF_RT1308_SSP_CODEC is used to encode the SSP port id for the amplifier. I rename it to SOF_AMPLIFIER_SSP.

File: sound/soc/intel/boards/Kconfig (hunk 5 above)
Comment (brentlu) [reply sound/soc/intel/boards/Kconfig:622]: Change the config name to SND_SOC_INTEL_SOF_RT1308_CS35L41_MACH. Also update the Makefile.

Comment (bardliao) [reply sound/soc/intel/boards/Kconfig:622]: Would it be better to call it something like SND_SOC_INTEL_SOF_SSP_AMP_MACH and sof_ssp_amp.c?

Comment (plbossart) [reply sound/soc/intel/boards/Kconfig:622]: agree with @bardliao, this becomes a template for solutions without a headset codec, so the naming should be generic,

**Review by bardliao — APPROVED** [review]

**Review by plbossart — APPROVED** [review]

**Pull request merged**